components/fxa-client/src/oauth.rs

```diff
@@ -172,25 +172,47 @@ impl FirefoxAccount {
         &self,
         auth_params: AuthorizationParameters,
     ) -> Result<String> {
-        // TODO: Add validation to check for which keys the client is allowed to request,
-        //       And if we have those keys stored locally.
-        //       Currently we pull ALL available keys that match the requested scopes, regardless
-        //       if the client is actually allowed to request them.
         let session_token = self.get_session_token()?;
+
+        // Validate request to ensure that the client is actually allowed to request
+        // the scopes they requested
+        let valid_scopes = self.client.scoped_key_data(
+            &self.state.config,
+            &session_token,
+            &auth_params.client_id,
+            &auth_params.scope,
+        )?;
+
+        if let Some(not_allowed_scope) = auth_params
+            .scope
+            .split_whitespace()
+            .find(|scope| !valid_scopes.contains_key(*scope))
+        {
+            return Err(ErrorKind::NotAllowedToRequestScope(not_allowed_scope.to_string()).into());
+        }
+
         let keys_jwe = if let Some(keys_jwk) = auth_params.keys_jwk {
             let keys_jwk = base64::decode_config(keys_jwk, base64::URL_SAFE_NO_PAD)?;
             let keys_jwk = String::from_utf8(keys_jwk)?;
-            let requested_scopes: HashSet<&str> = auth_params.scope.split_whitespace().collect();
-            let data: HashMap<String, ScopedKey> = self
+            let scoped_keys: HashMap<String, ScopedKey> = self
                 .state
                 .scoped_keys
                 .iter()
-                .filter(|(scope, _)| requested_scopes.contains(&(*scope).as_ref()))
+                .filter(|(scope, _)| valid_scopes.contains_key(*scope))
                 .map(|(key, value)| (key.clone(), value.clone()))
                 .collect();
-            let data = serde_json::to_string(&data)?;
+
+            // Return an error if we could not get the requested scopes from state
+            // Saves the client time instead of waiting for the server's response
+            let first_mismatch = scoped_keys
+                .iter()
+                .find(|(scope, _)| !valid_scopes.contains_key(*scope));
+            if let Some(scope) = first_mismatch {
+                return Err(ErrorKind::NoScopedKey(scope.0.clone()).into());
+            }
+            let scoped_keys = serde_json::to_string(&scoped_keys)?;
             let client_keys_flow = ScopedKeysFlow::with_random_key()?;
-            Some(client_keys_flow.encrypt_keys_jwe(&keys_jwk, data.as_bytes())?)
+            Some(client_keys_flow.encrypt_keys_jwe(&keys_jwk, scoped_keys.as_bytes())?)
```

Comment:
I'd like to see a couple of tests for correct behaviour here, especially given the security-sensitive nature of the checks you're doing. We should check that:

* If the server returns an error, the client receives an error.
* If the server returns a list of allowed scopes that is smaller than those requested by the client, then the client does not get the disallowed scopes.
* That the `NoScopedKey` error you're throwing above is actually throw when expected.

Reply:
Alrighty added a couple of tests in the integration tests workflow, although the position of the tests feels a little awkward 😅 

One thing I did notice however is that the server seems to return an error if we are requesting valid scopes that we are not allowed to request, and a bit of digging into the fxa server seems to align with that: [validation function in fxa](https://github.com/mozilla/fxa/blob/166d5871465334e1e2b4b237ad3f6cc00095455f/packages/fxa-auth-server/lib/oauth/grant.js#L96-L114)

Although with invalid scopes it does seem to filter them out, what I do in that case right now is error out.

Reply:
> One thing I did notice however is that the server seems to return an error if we are requesting valid scopes
> that we are not allowed to request, and a bit of digging into the fxa server seems to align with that

Gotcha. That makes sense I think, and as long as we propagate the error things will be fine. It's probably the better behaviour anyway rather than just pretending those scopes were not requested :-)

Reply:
> Alrighty added a couple of tests in the integration tests workflow, although the position of
> the tests feels a little awkward 😅

Heh, I see what you mean about being awkward (although FWIW I'm pleasantly surprised by how little code was required for each test).

Is this the sort of thing that might be a better fit for unittest-style tests, as part of `mod tests` further down in this file? I haven't looked into it in any detail, but it seems plausible that we could mock out `self.client.scoped_key_data` to test some of the edge-cases, without having to spin up a full-blown integration test. Thoughts?

Reply:
That does indeed work better and gets rid of the awkwardness! Thanks Ryan!